File: android/app/src/main/java/com/xirigo/ecommerce/core/designsystem/component/XGErrorView.kt
package com.xirigo.ecommerce.core.designsystem.component

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.ErrorOutline
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import com.xirigo.ecommerce.R
import com.xirigo.ecommerce.core.designsystem.theme.XGMotion
import com.xirigo.ecommerce.core.designsystem.theme.XGSpacing
import com.xirigo.ecommerce.core.designsystem.theme.XGTheme

/**
 * Full-screen error state with crossfade transition.
 *
 * Crossfades between [content] and the error layout (icon + message + retry button)
 * using [XGMotion.Crossfade.CONTENT_SWITCH] (200 ms). When [isError] is `true`,
 * the error layout is shown. When `false`, the [content] slot is shown.
 *
 * @param message Error message to display.
 * @param isError When `true`, crossfades to the error layout. When `false`, shows [content].
 * @param modifier Modifier applied to the outer container.
 * @param onRetry Optional retry callback. When non-null, a retry button is shown.
 * @param content The content to show when [isError] is `false`.
 */
@Composable
fun XGErrorView(
    message: String,
    isError: Boolean,
    modifier: Modifier = Modifier,
    onRetry: (() -> Unit)? = null,
    content: @Composable () -> Unit,
) {
    AnimatedContent(
        targetState = isError,
        modifier = modifier.fillMaxSize(),
        transitionSpec = {
            fadeIn(tween(XGMotion.Crossfade.CONTENT_SWITCH)) togetherWith
                fadeOut(tween(XGMotion.Crossfade.CONTENT_SWITCH))
        },
        label = "xgErrorViewCrossfade",
    ) { showError ->
        if (showError) {
            ErrorContent(
                message = message,
                onRetry = onRetry,
            )
        } else {
            content()
        }
    }
}

/** Full-screen error state with message and optional retry button (no crossfade). */
@Composable
fun XGErrorView(
    message: String,
    modifier: Modifier = Modifier,
    onRetry: (() -> Unit)? = null,
) {
    Box(
        modifier = modifier.fillMaxSize(),
        contentAlignment = Alignment.Center,
    ) {
        ErrorContent(
            message = message,
            onRetry = onRetry,
        )
    }
}

/** Internal error content layout: icon + message + optional retry button. */
@Composable
private fun ErrorContent(message: String, onRetry: (() -> Unit)?) {
    Box(
        modifier = Modifier.fillMaxSize(),
        contentAlignment = Alignment.Center,
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
        ) {
            Icon(
                imageVector = Icons.Outlined.ErrorOutline,
                contentDescription = null,
                modifier = Modifier.size(XGSpacing.XXL),
                tint = MaterialTheme.colorScheme.error,
            )

            Spacer(modifier = Modifier.height(XGSpacing.Base))

            Text(
                text = message,
                style = MaterialTheme.typography.bodyLarge,
                textAlign = TextAlign.Center,
                color = MaterialTheme.colorScheme.onSurface,
            )

            if (onRetry != null) {
                Spacer(modifier = Modifier.height(XGSpacing.Base))
                XGButton(
                    text = stringResource(R.string.common_retry_button),
                    onClick = onRetry,
                    style = XGButtonStyle.Primary,
                    fullWidth = false,
                )
            }
        }
    }
}

// region Preview

private const val PREVIEW_ERROR_MESSAGE = "Something went wrong"
private const val PREVIEW_CONTENT_TEXT = "Content loaded!"

@Preview(showBackground = true)
@Composable
private fun XGErrorViewPreview() {
    XGTheme {
        XGErrorView(
            message = PREVIEW_ERROR_MESSAGE,
            onRetry = {},
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun XGErrorViewNoRetryPreview() {
    XGTheme {
        XGErrorView(message = "No internet connection")
    }
}

@Preview(showBackground = true, name = "XGErrorView — Crossfade Error")
@Composable
private fun XGErrorViewCrossfadeErrorPreview() {
    XGTheme {
        XGErrorView(
            message = PREVIEW_ERROR_MESSAGE,
            isError = true,
            onRetry = {},
        ) {
            Text(
                text = PREVIEW_CONTENT_TEXT,
                style = MaterialTheme.typography.bodyLarge,
                modifier = Modifier.padding(XGSpacing.Base),
            )
        }
    }
}

@Preview(showBackground = true, name = "XGErrorView — Crossfade Content")
@Composable
private fun XGErrorViewCrossfadeContentPreview() {
    XGTheme {
        XGErrorView(
            message = PREVIEW_ERROR_MESSAGE,
            isError = false,
            onRetry = {},
        ) {
            Text(
                text = PREVIEW_CONTENT_TEXT,
                style = MaterialTheme.typography.bodyLarge,
                modifier = Modifier.padding(XGSpacing.Base),
            )
        }
    }
}

@Preview(showBackground = true, name = "XGErrorView — Interactive Toggle")
@Composable
private fun XGErrorViewInteractivePreview() {
    XGTheme {
        var isError by remember { mutableStateOf(false) }

        Column {
            XGErrorView(
                message = PREVIEW_ERROR_MESSAGE,
                isError = isError,
                modifier = Modifier.weight(1f),
                onRetry = { isError = false },
            ) {
                Text(
                    text = "Real content loaded!",
                    style = MaterialTheme.typography.bodyLarge,
                    modifier = Modifier.padding(XGSpacing.Base),
                )
            }

            androidx.compose.material3.Button(
                onClick = { isError = !isError },
                modifier = Modifier.padding(XGSpacing.Base),
            ) {
                Text(text = if (isError) "Show Content" else "Show Error")
            }
        }
    }
}

// endregion

File: ios/XiriGoEcommerce/Core/DesignSystem/Component/XGErrorView.swift
import SwiftUI

// MARK: - XGErrorView

/// Full-screen error state with crossfade transition.
///
/// Crossfades between content and the error layout (icon + message + retry button)
/// using ``XGMotion/Crossfade/contentSwitch`` (0.2s). When `isError` is `true`,
/// the error layout is shown. When `false`, the content slot is shown.
///
/// Usage with crossfade:
/// ```swift
/// XGErrorView(
///     message: "Something went wrong",
///     isError: viewModel.hasError,
///     onRetry: { viewModel.retry() }
/// ) {
///     ProductDetailContent(product: product)
/// }
/// ```
///
/// Usage without crossfade (backward compatible):
/// ```swift
/// XGErrorView(message: "Something went wrong", onRetry: { viewModel.retry() })
/// ```
struct XGErrorView<Content: View>: View {
    // MARK: - Lifecycle

    /// Creates an error view with crossfade transition.
    ///
    /// - Parameters:
    ///   - message: Error message to display.
    ///   - isError: When `true`, shows the error layout. When `false`, shows content.
    ///   - onRetry: Optional retry callback. When non-nil, a retry button is shown.
    ///   - content: Content to show when `isError` is `false`.
    init(
        message: String,
        isError: Bool,
        onRetry: (() -> Void)? = nil,
        @ViewBuilder content: () -> Content,
    ) {
        self.message = message
        self.isError = isError
        self.onRetry = onRetry
        self.content = content()
    }

    // MARK: - Internal

    var body: some View {
        Group {
            if isError {
                ErrorContent(message: message, onRetry: onRetry)
                    .transition(.opacity)
            } else {
                content
                    .transition(.opacity)
            }
        }
        .animation(
            .easeInOut(duration: XGMotion.Crossfade.contentSwitch),
            value: isError,
        )
    }

    // MARK: - Private

    private let message: String
    private let isError: Bool
    private let onRetry: (() -> Void)?
    private let content: Content
}

// MARK: - XGErrorView convenience (no crossfade)

extension XGErrorView where Content == EmptyView {
    /// Creates a static error view without crossfade (backward compatible).
    ///
    /// - Parameters:
    ///   - message: Error message to display.
    ///   - onRetry: Optional retry callback. When non-nil, a retry button is shown.
    init(
        message: String,
        onRetry: (() -> Void)? = nil,
    ) {
        self.message = message
        isError = true
        self.onRetry = onRetry
        content = EmptyView()
    }
}

// MARK: - ErrorContent

/// Internal error content layout: icon + message + optional retry button.
private struct ErrorContent: View {
    let message: String
    let onRetry: (() -> Void)?

    var body: some View {
        VStack(spacing: XGSpacing.base) {
            Image(systemName: "exclamationmark.circle")
                .font(.system(size: XGSpacing.IconSize.extraLarge))
                .foregroundStyle(XGColors.error)
                .accessibilityHidden(true)

            Text(message)
                .font(XGTypography.bodyLarge)
                .foregroundStyle(XGColors.onSurface)
                .multilineTextAlignment(.center)

            if let onRetry {
                XGButton(
                    String(localized: "common_retry_button"),
                    variant: .outlined,
                    fullWidth: false,
                    action: onRetry,
                )
            }
        }
        .padding(XGSpacing.base)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

// MARK: - Previews

#Preview("XGErrorView with Retry") {
    XGErrorView(
        message: String(localized: "common_error_generic"),
        onRetry: {},
    )
    .xgTheme()
}

#Preview("XGErrorView without Retry") {
    XGErrorView(
        message: "Network connection lost",
    )
    .xgTheme()
}

#Preview("XGErrorView — Crossfade Error") {
    XGErrorView(
        message: "Something went wrong",
        isError: true,
        onRetry: {},
        content: {
            Text(verbatim: "Content loaded!")
                .padding(XGSpacing.base)
        },
    )
    .xgTheme()
}

#Preview("XGErrorView — Crossfade Content") {
    XGErrorView(
        message: "Something went wrong",
        isError: false,
        onRetry: {},
        content: {
            Text(verbatim: "Content loaded!")
                .padding(XGSpacing.base)
        },
    )
    .xgTheme()
}

// MARK: - XGErrorViewInteractiveDemo

/// Interactive demo showing the crossfade transition between content and error.
private struct XGErrorViewInteractiveDemo: View {
    // MARK: - Internal

    var body: some View {
        VStack {
            XGErrorView(
                message: "Something went wrong",
                isError: isError,
                onRetry: { isError = false },
                content: {
                    Text(verbatim: "Real content loaded!")
                        .padding(XGSpacing.base)
                },
            )

            Button {
                isError.toggle()
            } label: {
                Text(verbatim: isError ? "Show Content" : "Show Error")
            }
            .padding(XGSpacing.base)
        }
    }

    // MARK: - Private

    @State private var isError = false
}

#Preview("XGErrorView — Interactive Toggle") {
    XGErrorViewInteractiveDemo()
        .xgTheme()
}
